### ISSUE TITLE
[bug] There is a probability that creating a new window will result in garbled folders in the software runtime directory, which is the data directory of webview

### ISSUE THREAD

**Issue by itisl2220** [tauri-apps/tauri#9501]
### Describe the bug

There is a probability that creating a new window will result in garbled folders in the software runtime directory, which is the data directory of webview
![image](https://github.com/tauri-apps/tauri/assets/32608524/98653e7c-bf91-4aa5-a499-0d35495de1b6)
![image](https://github.com/tauri-apps/tauri/assets/32608524/a4889a03-70c3-46da-b9d1-3eb07680595e)

### Full `tauri info` output

```text
[✔] Environment
    - OS: Windows 10.0.22631 X64
    ✔ WebView2: 123.0.2420.97
    ✔ MSVC:
        - Visual Studio Community 2022
        - Visual Studio 生成工具 2019
    ✔ rustc: 1.76.0 (07dca489a 2024-02-04)
    ✔ cargo: 1.76.0 (c84b36747 2024-01-18)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 18.15.0
    - pnpm: 8.10.0
    - yarn: 1.22.21
    - npm: 9.5.0

[-] Packages
    - tauri [RUST]: 2.0.0-beta.15
    - tauri-build [RUST]: 2.0.0-beta.12
    - wry [RUST]: 0.39.0
    - tao [RUST]: 0.27.0
    - tauri-cli [RUST]: 2.0.0-beta.12
    - @tauri-apps/api [NPM]: 2.0.0-beta.4
    - @tauri-apps/cli [NPM]: 2.0.0-beta.12

[-] App
    - build-type: bundle
    - CSP: unset
    - frontendDist: ../dist
    - devUrl: http://localhost:1420/
    - framework: Vue.js (Vue CLI)
    - bundler: Vite
```

**Comment by itisl2220** [comment]
![1713456144274](https://github.com/tauri-apps/tauri/assets/32608524/93e68fb4-03c6-474e-83da-3992a22ccb40)

**Comment by FabianLars** [comment]
Thanks for the report, i forgot to open an issue about this 😅 A tiny bit more context on discord: https://discord.com/channels/616186924390023171/1229300159607996437 (nothing of much value in there tbh)

**Comment by CrendKing** [comment]
Here is what I found:

The function of concern is [`create_environment()` from wry](https://github.com/tauri-apps/wry/blob/dev/src/webview2/mod.rs#L242).

Since random data appear in this issue, I first suspect the string is not null-terminated. However, I tested that `HSTRING::from` does append `\0` at the end.

Then I suspected use after free. So I added some logging of `data_directory`'s content. Specifically
```
diff --git a/src/webview2/mod.rs b/src/webview2/mod.rs
index 32dd2e2..be97c90 100644
--- a/src/webview2/mod.rs
+++ b/src/webview2/mod.rs
@@ -304,6 +303,7 @@ impl InnerWebView {
         );
         options.SetLanguage(PCWSTR::from_raw(lang.as_ptr()))?;

+        dbg!(std::slice::from_raw_parts(data_directory.unwrap().0, 50));
         CreateCoreWebView2EnvironmentWithOptions(
           PCWSTR::null(),
           data_directory.unwrap_or_else(PCWSTR::null),
@@ -314,11 +314,13 @@ impl InnerWebView {
       }),
       Box::new(move |error_code, environment| {
         error_code?;
+        unsafe { dbg!(std::slice::from_raw_parts(data_directory.unwrap().0, 50)); }
         tx.send(environment.ok_or_else(|| windows::core::Error::from(E_POINTER)))
           .map_err(|_| windows::core::Error::from(E_UNEXPECTED))
       }),
     )?;

+    unsafe { dbg!(std::slice::from_raw_parts(data_directory.unwrap().0, 50)); }
     rx.recv()?.map_err(Into::into)
   }
```

The problem happens randomly. Sometimes it is a corrupted data directory, sometimes it is the correct directory from `%APPDATA%`. When the problem happens, all 3 slices tend to have different data, none of them is correct. When the problem doesn't happen, the data right above `CreateCoreWebView2EnvironmentWithOptions()` would be correct, though the other two are likely to be wrong. Apparently something changed or corrupted the memory of the pointer.

There is something weird about the first `data_directory` Option at the top of the function. Here is a patch to seemingly fix the whole issue:
```
diff --git a/src/webview2/mod.rs b/src/webview2/mod.rs
index 32dd2e2..fc4cae1 100644
--- a/src/webview2/mod.rs
+++ b/src/webview2/mod.rs
@@ -248,7 +248,7 @@ impl InnerWebView {
       .as_deref()
       .and_then(|context| context.data_directory())
       .and_then(|path| path.to_str())
-      .map(String::from);
+      .map(HSTRING::from);

     // additional browser args
     let additional_browser_args = pl_attrs.additional_browser_args.unwrap_or_else(|| {
@@ -283,8 +283,7 @@ impl InnerWebView {
     let additional_browser_args = HSTRING::from(additional_browser_args);
     let additional_browser_args = PCWSTR::from_raw(additional_browser_args.as_ptr());

-    let data_directory = data_directory.map(HSTRING::from);
-    let data_directory = data_directory.map(|d| PCWSTR::from_raw(d.as_ptr()));
+    let data_directory_param = data_directory.clone().map(|d| PCWSTR::from_raw(d.as_ptr()));

     let (tx, rx) = mpsc::channel();
     CreateCoreWebView2EnvironmentCompletedHandler::wait_for_async_operation(
@@ -306,7 +305,7 @@ impl InnerWebView {

         CreateCoreWebView2EnvironmentWithOptions(
           PCWSTR::null(),
-          data_directory.unwrap_or_else(PCWSTR::null),
+          data_directory_param.unwrap_or_else(PCWSTR::null),
           &options,
           &environmentcreatedhandler,
         )
```
Basically adding a `.clone()` to the first Option solves the issue. I don't know why it works. Also, if we do
```
diff --git a/src/webview2/mod.rs b/src/webview2/mod.rs
index fc4cae1..25d924d 100644
--- a/src/webview2/mod.rs
+++ b/src/webview2/mod.rs
@@ -303,6 +303,7 @@ impl InnerWebView {
         );
         options.SetLanguage(PCWSTR::from_raw(lang.as_ptr()))?;

+        dbg!(data_directory);
         CreateCoreWebView2EnvironmentWithOptions(
           PCWSTR::null(),
           data_directory_param.unwrap_or_else(PCWSTR::null),
```
The problem comes back.

Also, shouldn't `additional_browser_args` suffers the same issue, since it has the same type?

**Comment by amrbashir** [comment]
@CrendKing Thanks for you help, and very nice finding.

> Basically adding a `.clone()` to the first Option solves the issue. I don't know why it works.

I suspect because `HSTRING` is a reference-counted type and so when you clone it, you increase the strong count and it is not dropped eagerly before `CreateCoreWebView2EnvironmentWithOptions` returns.

I have implemented your solution in https://github.com/tauri-apps/wry/pull/1221 and attributed you in the commit.
